### ISSUE TITLE
Kernel crash on Windows

### ISSUE THREAD

**Issue by borisdjartov** [xcsf-dev/xcsf#127]
Hi guys,

thank you for putting in so much time in this project. I am just encountering issues when i try to train the model as it crasehs my Python kernel. Sometimes it seems to magically work but i am not sure what i am doing wrong.  Bellow i have some screenshots from my jupyter notebook where the problems occures.

Getting the data ready
![image](https://github.com/xcsf-dev/xcsf/assets/44361789/b9483d07-d0e0-4d40-af46-27c00678132c)
The model defined for a supervised learning task
![image](https://github.com/xcsf-dev/xcsf/assets/44361789/c1e4a2f3-83dd-4d8e-bb22-06ba557afbd8)
The part where the kernel crashes! When i use the .fit method
![image](https://github.com/xcsf-dev/xcsf/assets/44361789/4937775d-87d6-4580-b54e-530bb0f827ad)

I have gone over the inistilation requirements and everything seems to be installed properly, however it does not work. 

I am running it on Windows 11 python version 3.10.10

Any help would be greatly apreciated.

Kind regards,
Boris

**Comment by rpreen** [comment]
Can you try running it from a script rather than a notebook? Jupyter notebooks seem to hide the stdout and if it's crashing it should be writing an error to the terminal (command prompt) that you would be able to see running it: 
```
$ python test.py
```

It's hard to see what's wrong from the above, but I suspect that it's something to do with the shape of the data. Can you print the shape of what's being fed to XCSF? Something like:
```python
print(f"X_train shape = {np.shape(X_train)}")
print(f"y_train shape = {np.shape(y_train)}")
print(f"X_val shape = {np.shape(X_val)}")
print(f"y_val shape = {np.shape(y_val)}")
print(f"X_test shape = {np.shape(X_test)}")
print(f"y_test shape = {np.shape(y_test)}")
```

Also, before caling `fit()`, display the XCSF configuration to make sure it's exactly what you intend and every hyperparameter has been set correctly:
```python
print(json.dumps(xcs.internal_params(), indent=4))
```

**Comment by rpreen** [comment]
Also, I don't think using least squares predictions for one-hot encoding via log loss will work. I have only used the neural networks with softmax outputs for log loss. For the log loss error calculation, the sum of a classifier's predictions must be 1, which wont necessarily be true for NLMS and that might cause problems.

**Comment by borisdjartov** [comment]
Hi thank you for the fast reply.

I tried running it from a script and it just does nothing it kinda stops on its own. No error message or anything. Here are the x and y shapes.
![image](https://github.com/xcsf-dev/xcsf/assets/44361789/09e8396a-e370-4c12-821e-f77193acfc78)


Here is the printed xcf
{
    "version": "1.4.4",
    "x_dim": 21,
    "y_dim": 3,
    "n_actions": 1,
    "omp_num_threads": 4,
    "random_state": 10,
    "population_file": "",
    "pop_init": true,
    "max_trials": 5000,
    "perf_trials": 500,
    "pop_size": 100,
    "loss_func": "log",
    "set_subsumption": false,
    "theta_sub": 100,
    "e0": 0.01,
    "alpha": 1,
    "nu": 5,
    "beta": 0.05,
    "delta": 0.1,
    "theta_del": 200,
    "init_fitness": 0.01,
    "init_error": 0,
    "m_probation": 10000,
    "stateful": true,
    "compaction": false,
    "ea": {
        "select_type": "tournament",
        "select_size": 0.4,
        "theta_ea": 50,
        "lambda": 2,
        "p_crossover": 0.7,
        "err_reduc": 1,
        "fit_reduc": 0.1,
        "subsumption": true,
        "pred_reset": false
    },
    "condition": {
        "type": "hyperrectangle_ubr",
        "args": {
            "eta": 0,
            "min": 0,
            "max": 1,
            "spread_min": 1
        }
    },
    "prediction": {
        "type": "nlms_linear",
        "args": {
            "x0": 1,
            "eta": 1,
            "evolve_eta": true,
            "eta_min": 0.0001
        }
    }
}

The error 
![image](https://github.com/xcsf-dev/xcsf/assets/44361789/ad2eb59b-ea2a-421e-a224-65ba6b2a0352)

The log 
![image](https://github.com/xcsf-dev/xcsf/assets/44361789/7569a7ff-aa1d-4e94-8c71-48ce27e0bbed)

Additionally i also tried to just run it for an older dataset that it worked on and now it doesn't work on it anymore and the kernel crashes. I will try to run the neural network one but i am not so sure that it will work. 
Thank you again for the fast reply.

Kind regards,
Boris

**Comment by rpreen** [comment]
Does the `example_classification.ipynb` work out of the box or does that crash too?

**Comment by rpreen** [comment]
In the `TERMINAL` tab on your notebook does that show the stdout?

**Comment by rpreen** [comment]
I'm noticing some weird stuff trying to run the examples in a Windows VM now... some of the examples (rmux and maze) work but the regression and classification are exiting silently after the call to fit. Everything seems to work fine on Linux however. I'll see if I can find out why...

**Comment by borisdjartov** [comment]
Thank you for the reply.

So it might just be a weird windows problem. I will try and run it on a linux machine. The classification notebook also has the same problem.

**Comment by rpreen** [comment]
There seems to be a strange problem with `std::ostringstream` on Windows which causes it to crash (it's used to build the performance status for displaying in the fit function). So turning `verbose=False` will avoid the crash, although not a solution. Oddly it works fine if XCSF is compiled as debug, but this will run very slowly, so again not a solution. This is all really strange because I'm sure it was working fine on Windows when I tested a while ago.

**Comment by borisdjartov** [comment]
Hi i tried setting verbose to False but it still didn't work 😔.

**Comment by rpreen** [comment]
Did you set it to False in both fit and the callback?

**Comment by borisdjartov** [comment]
No i just did and it seems like it works. It did not crash and i tried to usu the trained model to predict something altough i got a strange answer back but it seems like it did run.

Thank you so much for the help i really apreciate it. Not too much work being done on LCS.

**Comment by rpreen** [comment]
I suggest using the neural nets for the predictions if using one-hot encoding for classification and log loss; you could always just have a single connected layer with linear activations followed by a softmax, which would be very similar to having the linear least squares weight vectors.

Hopefully I can fix this bug screwing up the Windows version soon -- I'm really confused why it was working before but not now - and even the older compiled versions when installed from pip seem to also crash now when they didn't before.

**Comment by borisdjartov** [comment]
Yes i was also very confused I had used this before and it did work! For the output i just passed it through a softmax so i can get the probabilities and then assign the class to the highest one. It seems to work but i am analyzing the data now.

Thank you so much for the help and i hope the problem can be fixed soon.

**Comment by rpreen** [comment]
@borisdjartov I've built and uploaded v1.4.5 to pip. 

Please can you install that version and let me know if it fixed the problem for you?

**Comment by borisdjartov** [comment]
I just chekced it out and it seems to work now, thank you very much!
